src/easylink/configuration.py

```diff
@@ -45,42 +85,54 @@ class Config:
 
     def __init__(
         self,
-        pipeline_specification: Union[str, Path],
-        input_data: Union[str, Path],
-        computing_environment: Optional[Union[str, Path]],
-        results_dir: Union[str, Path],
+        config_params: Dict,
     ):
-        # Handle pipeline specification
-        self.pipeline = load_yaml(pipeline_specification)
-        self._requires_spark = self._determine_if_spark_is_required(self.pipeline)
-        # Handle input data specification
-        self.input_data = self._load_input_data_paths(input_data)
-        # Handle environment specification
-        self.environment = self._load_computing_environment(computing_environment)
-        # Store path to results directory
-        self.results_dir = Path(results_dir)
-
-        # Extract environment attributes and assign defaults as necessary
-        self.computing_environment = self._get_required_attribute(
-            self.environment, "computing_environment"
-        )
-        self.container_engine = self._get_required_attribute(
-            self.environment, "container_engine"
-        )
-        self.slurm = self.environment.get("slurm", {})  # no defaults for slurm
-        self.implementation_resources = self._get_implementation_resource_requests(
-            self.environment
-        )
-        self.spark = self._get_spark_requests(self.environment, self._requires_spark)
+        super().__init__(layers=["initial_data", "default", "user_configured"])
+        self.update(DEFAULT_ENVIRONMENT, layer="default")
+        self.update(config_params, layer="user_configured")
+        for step in self.pipeline["steps"]:
+            self.update(
+                {"pipeline": {"steps": {step: {"implementation": {"configuration": {}}}}}},
+                layer="default",
+            )
+        if self._determine_if_spark_is_required(self.pipeline):
+            self.update({"environment": {"spark": SPARK_DEFAULTS}}, layer="default")
+        else:
+            # Is there a less hacky way to remove this?
```

Comment:
How did it get generated in the first place if no implementations required it? That aside, I don't think there's a better way.

Reply:
This is to cover the case that a user specifies a spark key in the `environment` yaml but doesn't use any spark-required implementations. I guess it is fine to also just leave this as being vestigial, but the previous behavior was to omit them from the config entirely, so I replicated that.